invisible_cities/cities/detsim.py

```diff
@@ -98,39 +98,29 @@ def create_s2_waveform(xs, ys, ts, phs, tmin, buffer_length):
     return create_s2_waveform
 
 
-def bin_edges_getter(wf_pmt_bin_width, wf_sipm_bin_width):
+def bin_edges_getter(pmt_width, sipm_width):
     """
     Auxiliar function that returns the waveform bin edges
     """
     def get_bin_edges(pmt_wfs, sipm_wfs):
-        pmt_bins  = np.arange(0, pmt_wfs .shape[1]) * wf_pmt_bin_width
-        sipm_bins = np.arange(0, sipm_wfs.shape[1]) * wf_sipm_bin_width
+        pmt_bins  = np.arange(0, pmt_wfs .shape[1]) * pmt_width
+        sipm_bins = np.arange(0, sipm_wfs.shape[1]) * sipm_width
         return pmt_bins, sipm_bins
     return get_bin_edges
 
 
 @city
 def detsim(files_in, file_out, event_range, print_mod, compression,
            detector_db, run_number, s1_lighttable, s2_lighttable, sipm_psf,
-           buffer_parameters  = dict(),
-           physics_parameters = dict()):
+           buffer_params  = dict(),
```

Comment:
Eeeeek! Don't use mutable objects as default arguments!

(Do you know how to deal with this, or shall I explain?)

Reply:
Please explain. Is not the same case in [dorothea](https://github.com/next-exp/IC/blob/master/invisible_cities/cities/dorothea.py), [penthesilea](https://github.com/next-exp/IC/blob/master/invisible_cities/cities/penthesilea.py) and [esmeralda](https://github.com/next-exp/IC/blob/master/invisible_cities/cities/esmeralda.py)?

Reply:
The fundamental problem is that (in Python) default arguments are evaluated only once, at the time the function is defined. Consequently, the same *single object* is shared by all calls to the function which do not provide an overriding argument. If this object is mutated in any of those calls, then all subsequent calls use a different default value from the one visible in the function's parameter list. When this happens, the resulting observations (if you spot them; you might not, and then you're producing incorrect results) tend to be so freakishly weird, and waste so much debugging time, that we simply want to make sure that we never get to that stage. So, we studiously avoid *ever* using mutables as default arguments, directly.

Try this simple experiment to see it in action:

``` python
def foo(x, a=[]):
    a.append(x)
    return max(x, *a)

foo(666)
foo(1)
```

It's a subtle manifestation of the scourge of mutable shared state.

A simple solution would be to use a frozen dict. Unfortunately a PEP proposing frozen dicts was rejected some years ago, leaving us with the bizarre situation that `frozenset` is a builtin, while its dictionary brother was not allowed in.

This leaves two obvious options:

1. Use the frozen dict package. Web search left as an exercise. I've no experience with it, so I can't vouch for it.

2. Use the standard idiom for dealing with mutable default arguments:

   ```python
   def fn(parameter=None):
       if parameter is None:
           parameter = <create *NEW INSTANCE* of mutable default value here>
   ```

   which is a bit annoying, but always works.
   
> Is not the same case in dorothea, penthesilea and esmeralda?

If so, then it should be fixed there too!

Not necessarily in this PR. If there is someone new to Next/IC/Git/PRs/our-workflow, nervously waiting in the wings, then fixing this problem in those other places could be an excellent introduction to submitting a PR in our environment.

Reply:
Oh ...

> Is not the same case in dorothea, penthesilea and esmeralda?

... let this act as an illustration of why the practice of copy pasting stuff without understanding what it means and why it is done the way it is, has significant costs and is thoroughly discouraged.

Invest time and effort in understanding things deeply, and it will pay dividends in the long term.

Reply:
But this is a city, i.e. this function will never be called twice in the same program. At least it is not meant to. Isn't that a good motivation to reduce the noise introduced by initializing them to `None`? On the other hand, I don't know why these are the only parameters that are initialized to "nothing" as if they were optional, aren't they essential to the running of the city?

Reply:
Yes, they are essential. Those parameters being an empty dictionary as they are by default won't work. They config file must contains all the parameters defined.

Reply:
> this function will never be called twice in the same program. 

You hope ...

> At least it is not meant to.

... there you go: things than are not meant to happen, tend to have a habit of happening anyway, every now an then at least.

> Isn't that a good motivation to reduce the noise introduced by initializing them to None?

It may be.

A competent Python programmer should get very nervous at seeing mutable default values, and will then have to waste time wondering about the safety, or fixing the (non-)issue. So, if it stays, it must be accompanied by comments as to why it is safe (which the reader may, or may not, trust) ... at which point it might be less noisy simply to do the `None` trick.

Reply:
> Those parameters being an empty dictionary as they are by default won't work. 

So the default value defined in the function must *always* be overridden for the function to stand a chance of working?

That's so wrong!

Why not turn them into keyword-only args?

Reply:
> Why not turn them into keyword-only args?

... in fact, IIRC, it would simply make sense to make all parameters of all cities keyword-only.

But, even without that, just remove the default values, and everything should work, shouldn't it?

Reply:
This is something inherent to all cities in IC. All arguments are actually handled as keyword arguments by the `city` decorator (IIRC). We could turn them into keyword-only arguments, and maybe we should as I don't think it is of any use to have positional arguments. Somehow it bothers me to have that `*` at the beginning of the function signature, but it is definitely something irrelevant for the decision.

Reply:
My hypotheses are

1. Adding `*` to the front of each city's signature should break nothing.
2. Removing all `= dict()` from all cities' signatures should break nothing.

I'm pretty confident on the first: that's by design of the whole liquid city thing. If anything breaks, it will be because, someone, somewhere is abusing something. Adding the `*` reduces the chances of someone abusing something in the future.

I'm less confident on the second: these are lower-level details so I'd have to look into the details case-by-case.

I propose removing the `= dict()`s in this city, and running all the tests. If that works, the obvious followups are the two points listed above, in reverse order.

Reply:
I agree. The first point will work for sure. The second one might break some random tests that don't care about the dict-like arguments to run, but it won't break any realistic usage of the city.